From fcb3b108ab88a5e7dc81c041f98c79622d88143c Mon Sep 17 00:00:00 2001
From: "Rafael G. Martins" <rafael@rafaelmartins.eng.br>
Date: Sun, 11 Oct 2015 18:13:34 -0300
Subject: [PATCH] git-shell: create hooks/ dir if needed

---
 blogc-git-receiver.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/blogc-git-receiver.c b/blogc-git-receiver.c
index a34e194..d575c79 100644
--- a/blogc-git-receiver.c
+++ b/blogc-git-receiver.c
@@ -218,6 +218,17 @@ git_shell(int argc, char *argv[])
         goto cleanup;
     }
 
+    if (0 != access("hooks", F_OK)) {
+        // openwrt git package won't install git templates, then the git
+        // repositories created with it won't have the hooks/ directory.
+        if (0 != mkdir("hooks", 0777)) {  // mkdir honors umask for us.
+            fprintf(stderr, "error: failed to create directory (%s/%s/hooks): "
+                "%s\n", home, repo, strerror(errno));
+            rv = 1;
+            goto cleanup;
+        }
+    }
+
     if (0 != chdir("hooks")) {
         fprintf(stderr, "error: failed to chdir (%s/%s/hooks): %s\n", home,
             repo, strerror(errno));
-- 
2.5.0

